builtin/providers/aws/resource_aws_instance.go

```diff
@@ -30,9 +30,17 @@ func resourceAwsInstance() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"ami": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"ami_name"},
+			},
+
+			"ami_name": &schema.Schema{
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"ami"},
```

Comment:
When `ami` is required, any `plan` will fail if it's not specified. With this change and both `ami` attributes being optional, users won't get a validation error but instead a Terraform failure with an unknown amount of infrastructure setup.